provider/openstack/cinder.go

```diff
@@ -0,0 +1,443 @@
+
+	resp, err := ga.cinder.GetVolumesSimple()
+	if err != nil {
+		return nil, err
+	}
+
+	vols := make([]storage.Volume, len(resp.Volumes))
+	for idx, vol := range resp.Volumes {
+		vols[idx] = cinderToJujuVolume(names.VolumeTag{}, &vol)
+	}
+
+	return vols, nil
+}
+
+func (ga *gooseAdapter) AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error) {
+	attachment, err := ga.nova.AttachVolume(serverId, volumeId, mountPoint)
+	if err != nil {
+		return storage.VolumeAttachment{}, err
+	}
+	return novaToJujuVolumeAttachment(attachment), nil
+}
+
+func (ga *gooseAdapter) VolumeStatusNotifier(volId, status string, numAttempts int, waitDur time.Duration) <-chan error {
+	return ga.cinder.VolumeStatusNotifier(volId, status, numAttempts, waitDur)
+}
+
+func (ga *gooseAdapter) ListVolumeAttachments(serverId string) ([]storage.VolumeAttachment, error) {
+	novaVolAttachments, err := ga.nova.ListVolumeAttachments(serverId)
+	if err != nil {
+		return nil, err
+	}
+
+	volumes := make([]storage.VolumeAttachment, len(novaVolAttachments))
+	for volIdx, volAttach := range novaVolAttachments {
+		volumes[volIdx] = novaToJujuVolumeAttachment(&volAttach)
+	}
+
+	return volumes, nil
+}
+
+func (ga *gooseAdapter) DetachVolume(serverId, attachmentId string) error {
+	return ga.nova.DetachVolume(serverId, attachmentId)
+}
+
+const (
+	openstackCfgKeyUrl        = "auth-url"
+	openstackCfgKeyUser       = "username"
+	openstackCfgKeyPassword   = "password"
+	openstackCfgKeyAccessKey  = "access-key"
+	openstackCfgKeySecretKey  = "secret-key"
+	openstackCfgKeyRegion     = "region"
+	openstackCfgKeyTenantName = "tenant-name"
+)
+
+// func NewOpenstackStorageConfig(url, user, secrets, region, tenantName string) (*storage.Config, error) {
+
+// 	return storage.NewConfig(
+// 		"cinder",
+// 		CinderProviderType,
+// 		map[string]interface{}{
+// 			openstackCfgKeyUrl:        url,
+// 			openstackCfgKeyUser:       user,
+// 			openstackCfgKeySecrets:    secrets,
+// 			openstackCfgKeyRegion:     region,
+// 			openstackCfgKeyTenantName: tenantName,
+// 		},
+// 	)
+// }
+
+func cinderToJujuVolume(tag names.VolumeTag, volume *cinder.Volume) storage.Volume {
+	var size uint64
+	size = uint64(int64(volume.Size * 1024))
+	return storage.Volume{
+		VolumeId: volume.ID,
+		Size:     size,
+		Tag:      tag,
+	}
+}
+
+func novaToJujuVolumeAttachment(volAttach *nova.VolumeAttachment) storage.VolumeAttachment {
+	return storage.VolumeAttachment{
+		Volume:     names.NewVolumeTag(volAttach.VolumeId),
+		Machine:    names.NewMachineTag(volAttach.ServerId),
+		DeviceName: volAttach.Device,
+	}
+}
+
+func detachVolume(instanceId, volId string, volAttachments []storage.VolumeAttachment, openstackClient OpenstackAdapter) error {
+
+	attachment := findAttachment(volId, volAttachments)
+	if attachment == nil {
+		return nil
+	}
+
+	return openstackClient.DetachVolume(instanceId, attachment.Volume.String())
```

Comment:
attachment.Volume is the wrong thing to be using here. You should be using the attachment ID from OpenStack, which is not recorded in storage.VolumeAttachment.

I think you should change the adapter so that it only deals with nova structures. I guess it wouldn't be an adapter at all, just an interface that implements the volume-related bits of Nova and Cinder.